### PR TITLE
Refactor initialization code and update typical init

### PR DESCRIPTION
**Pull request by Crphang**
Things Left To Do:

- [X] Refactor based on comments on the architecture
- [X] Update the `default` theme to actually reflect what is required in #861 
- [X] Add defensive code to validate given theme
- [x] Refactor Tests. This portion is rather major. A design doc will be updated in this thread on how unittests are refactored

**What is the purpose of this pull request? (put "X" next to an item, remove the rest)**

- [X] Enhancement to an existing feature

Fixes: #861 
Related to: #519

**What is the rationale for this request?**

This PR updates current initialization to illustrate a more complete set of features. On top of that, it also addresses the issue that editing our scaffold markbind or providing more initialization options is too complex in our current codebase which uses const hardcoded defaults.

**What changes did you make? (Give an overview)**

There were two main architecture that I explored in this PR on how to refactor this initialization code:

Copying the entire theme:
image

Creating all the files and directories, and copying from the theme if file exist for the theme:
image

**Provide some example code that this change will affect:**

The code below illustrates the change in Site.initSite based on the first architecture.

```js
/**
 * Static method for initializing a markbind site.
 * Generate the site.json and an index.md file.
 *
 * @param rootPath
 */
Site.initSite = function (rootPath, theme) {
  const themePath = path.join(__dirname, TEMPLATE_ROOT_FOLDER_NAME, theme);

  return new Promise((resolve, reject) => {
    fs.accessAsync(rootPath)
      .catch(() => fs.copy(themePath, rootPath))
      .then(resolve)
      .catch(reject);
  });
};
```

**Is there anything you'd like reviewers to focus on?**

Design | Pros | Cons
--- | --- | ---
*First Design which copies everything* | WYSIWYG nature can make it easy to work with for maintainers to understand the code of initialization because it is simple | The codebase will have many repetition as we need to provide empty directories for all themes/templates even if it is empty
*Second Design which creates directories first and only copies required or provided files* | Only need to provide the files that is required. <br> When adding a new directory, we don't have to go into each of those themes to add those empty directory if those themes don't require. | Similar to current codebase, we have to go through the entire list of file generation to understand what files are created.

This commit illustrates the first design, but I am more *for* the second design which is basically adding 1 additional line to each promise blocks of the previous code. Would appreciate some discussion here.

Example second design:

```js
function createAndCopyFileFromTheme(generateFileOrDirectory, objectPath, themePath) {
  generateFileOrDirectory
  .then(() => {
    if (!fs.existsSync(themePath)) {
      return Promise.resolve();
    }
    return fs.copyAsync(themePath, objectPath)
  })
  .catch(() => {
    return Promise.reject();
  })
}

Site.initSite = function (rootPath, theme) {
  const themePath = path.join(__dirname, TEMPLATE_ROOT_FOLDER_NAME, theme);
  const boilerplatePath = path.join(rootPath, BOILERPLATE_FOLDER_NAME);

  return new Promise((resolve, reject) => {
    fs.accessAsync(boilerplatePath)
      .catch(() => {
        if (fs.existsSync(boilerplatePath)) {
          return Promise.resolve();
        }
        const boilerplateThemePath = path.join(themePath, BOILERPLATE_FOLDER_NAME);

        return createAndCopyFileFromTheme(fs.mkdirp(boilerplatePath), boilerplatePath, boilerplateThemePath);
      })
      .then(resolve)
      .catch(reject);
  });
}
```

### CODE REVIEW

**Comment by Crphang** [comment]
Request for comments 

cc: @damithc

**Comment by damithc** [comment]
> cc: @damithc

As this is an internal design improvement, I'm not in a position to say much. I'll wait for senior devs to weigh in.

**Comment by yamgent** [comment]
I am leaning towards the second design, although I feel that both designs will start to show their flaws when we need to rename existing directories.

> **Second Design Pros:**
> Only need to provide the files that is required. 
> When adding a new directory, we don't have to go into each of those themes to add those empty directory if those themes don't require.

Actually, is there strictly a need to follow a directory structure for the template in the first place? I feel that this requirement will make it harder to maintain as we now have to ensure that both the code and the template directory structure are in sync with each other, which isn't desirable either.

**Comment by Crphang** [comment]
@yamgent very good points raised there, definitely there are some restrictions on both designs. 

On those points, I think a mixture of first and second design can provide both flexibility and proper safeguards. I treat the `code` directories as a form of Interface where all templates in the future need, these directories should be minimal. If these `code` directories need to be renamed or changed, the impact will be minimized.

We can then follow the first design to initialize the site if those basic directories exist (passes validation) by copying the entire template files.

What do you think?

**Comment by yamgent** [comment]
> On those points, I think a mixture of first and second design can provide both flexibility and proper safeguards. I treat the `code` directories as a form of Interface where all templates in the future need, these directories should be minimal. If these `code` directories need to be renamed or changed, the impact will be minimized.
> 
> We can then follow the first design to initialize the site if those basic directories exist (passes validation) by copying the entire template files.
> 
> What do you think?

Sure, let's evolve the design to that direction and see how it goes.

**Comment by Crphang** [comment]
Test changes:

1. Included validation in unit test to show test for validation.
1. Leverage more on MemFS to perform template mocking.

Previous  design:
image

Current  design:

image

Main differences lies in Site initialization process. Previous site initialization does not require user input, so the unit test was resembling an integration test. Current site initialization requires user input and the output depends on the template that was chosen. Since this is a Unit Test, user input does not matter for us (should be covered by integration tests for future templates added), we mock the template input accordingly.

**Comment by Crphang** [comment]
@yamgent wondering whether you have any insight on the failing test. It is passing locally for me, seems it could be because of the node_modules cache.

**Comment by yamgent** [comment]
> @yamgent wondering whether you have any insight on the failing test. It is passing locally for me, seems it could be because of the node_modules cache.

Were you running npm 5.8.0 when you update package-lock.json? I recall the newer version of npm handling versions differently, which may cause certain dependencies to be the wrong versions and could result in different test outputs generated.

**Comment by Crphang** [comment]
Ready for review. Thanks for your help @yamgent so far! I reverted to node 8 and npm 6 to get this working. Definitely some issue dependency issues there when using node 10 and upgrading our dependencies. It will cause deploy/netlify stage to fail I would attempt to reproduce in a fresh issue.

The architecture and design are stated in previous comments on how I approached this issue. Hopefully that will help in reviewing. I adopted open-close principal in allowing for developers to create new initialization templates, and also this design made it easier to integrate with remote templates in the future.

File: package-lock.json [hunk 1]
```diff
@@ -9587,4 +9587,4 @@
       }
     }
   }
-}
+}
```
Comment (yamgent) [inline package-lock.json]: This change should not happen.

File: src/Site.js [hunk 2]
```diff
@@ -231,119 +153,17 @@ function setExtension(filename, ext) {
...
-  const headFolderPath = path.join(rootPath, HEAD_FOLDER_PATH);
-  const indexPath = path.join(rootPath, INDEX_MARKDOWN_FILE);
-  const siteNavPath = path.join(rootPath, SITE_NAV_PATH);
-  const siteLayoutPath = path.join(rootPath, LAYOUT_FOLDER_PATH);
-  const siteLayoutDefaultPath = path.join(siteLayoutPath, LAYOUT_DEFAULT_NAME);
-  const siteDefaultLayoutScriptsPath = path.join(siteLayoutDefaultPath, LAYOUT_SCRIPTS_PATH);
-  const sitePluginPath = path.join(rootPath, PROJECT_PLUGIN_FOLDER_NAME);
-  const userDefinedVariablesPath = path.join(rootPath, USER_VARIABLES_PATH);
-  // TODO: log the generate info
+Site.initSite = function (rootPath, templatePath = 'default') {
```
Comment (yamgent) [inline src/Site.js]: The default value is not necessary since the caller in index.js already supplies the default value.

File: src/template/template.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    }
+  }
+
+  return true;
+}
+
+function generateSiteWithTemplate(rootPath, templatePath) {
+  return new Promise((resolve, reject) => {
+    fs.accessAsync(rootPath)
+      .catch(() => fs.emptydirSync(rootPath))
```
Comment (yamgent) [inline src/template/template.js]: I don't think it is appropriate to wipe the rootPath directory should generation fail, it may already have some existing content before MarkBind was invoked, and it is not a good idea to delete those files.

File: src/template/template.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+      .then(() => fs.copyAsync(templatePath, rootPath))
+      .then(() => {
+        resolve();
+      })
```
Comment (yamgent) [inline src/template/template.js]: The three lines above can be simplified to:

```js
        .then(resolve)
```

Same for the other places in the code.

File: src/template/default/site.json [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+      "glob": "**/index.md"
+    },
+    {
+      "glob": "**/*.+(md|mbd)"
+    }
+  ],
+  "deploy": {
+    "message": "Site Update."
+  }
+}
```
Comment (yamgent) [inline src/template/default/site.json]: Add newline

File: src/template/template.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    throw new Error('Template validation failed. Required files does not exist');
+  }
+
+  return new Promise((resolve, reject) => {
+    generateSiteWithTemplate(this.rootPath, templatePath)
+      .catch((err) => {
+        reject(err);
+      })
+      .then(() => resolve())
+      .catch(() => reject());
```
Comment (yamgent) [inline src/template/template.js]: Can be simplified to:

```js
        .catch(reject);
```

File: src/template/template.js [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  return new Promise((resolve, reject) => {
+    fs.accessAsync(rootPath)
+      .catch(() => fs.emptydirSync(rootPath))
+      .then(() => fs.copyAsync(templatePath, rootPath))
+      .then(() => {
+        resolve();
+      })
+      .catch((err) => {
+        reject(err);
+      });
```
Comment (yamgent) [inline src/template/template.js]: Does `.catch(reject)` not work for this case? I don't think we need to explicitly specify `err` unless we want to use the value (e.g. if we want to log the error message with the logger).

File: test/functional/test_site/expected/testImportVariables.html [hunk 8]
```diff
@@ -62,6 +62,19 @@ <h2 id="trying-to-access-an-imported-variable-via-namespace">Trying to access an
...
+<script>
+  window.dataLayer = window.dataLayer || [];
+
+  function gtag() {
+    dataLayer.push(arguments);
+  }
+  gtag('js', new Date());
+
+  gtag('config', 'TRACKING-ID');
+</script>
```
Comment (yamgent) [inline test/functional/test_site/expected/testImportVariables.html]: These changes seem to be unrelated to the PR, I wonder how come our tests didn't fail. :O

Comment (Crphang) [reply test/functional/test_site/expected/testImportVariables.html]: Shouldn't this be added as part of previous PR 🤔. Why didnt previous PR fail if it does not have this.

File: src/template/template.js (hunk 6 above)
Comment (Crphang) [reply src/template/template.js]: Hmm i'll change

File: src/template/template.js (hunk 3 above)
Comment (Crphang) [reply src/template/template.js]: I thought that was the behaviour, it creates the directory if accessAsync fails as it assumes that the directory does not exist.

Comment (yamgent) [reply src/template/template.js]: Perhaps `fs.mkdirSync()` would be more appropriate for implementing that behaviour instead?

File: test/functional/test_site/expected/testImportVariables.html (hunk 8 above)
Comment (yamgent) [reply test/functional/test_site/expected/testImportVariables.html]: I think it is because another separate PR was also merged together with your previous PR, and the separate PR added in some new files. Never mind, let's put the fix in this PR, but as a standalone commit.

**Comment by Crphang** [comment]
Made changes accordingly @yamgent. Thanks for helping to review!

I was wondering where would be most appropriate to place `page.ejs` since `template` dir is being hijacked :laughing:.

Based on my understanding, our current logic doesn't allow multiple forms of base template which means the current structure of `page.ejs` is quite coupled with our design. In that direction of thought, I would think that placing it at top level together with Page.js and Site.js would be appropriate. To limit the scope of this PR, I feel that it would be better to refactor this in a separate PR. What do you think?

**Comment by Crphang** [comment]
Great looking out there!

Added a fix and a unit test to check that behavior

File: src/util/fsUtil.js [hunk 9]
```diff
@@ -30,4 +52,6 @@ module.exports = {
     path.dirname(filePathWithExt),
     path.basename(filePathWithExt, path.extname(filePathWithExt)),
   ),
+
+  copySyncWithOptions,
```
Comment (yamgent) [inline src/util/fsUtil.js]: Can just put the code for `copySyncWithOptions()` here directly, since other methods also do it the same way for this file, let's try to be consistent.

**Comment by Crphang** [comment]
Changed according to comments. Ready for review @yamgent. Thanks a lot!

**Comment by yamgent** [comment]
> You might want to rename the first commit to follow the same as the PR title.

Apologies for being unclear, when I stated this, I meant removing the "[861-update-init-command]" part, so the title should only be "Refactor initialization code and update typical init". We don't put branch names in commit titles here. :P

Otherwise this looks pretty good. :+1:

**Comment by Crphang** [comment]
Updated the commit name. Also merged the third commit into the first. Thanks a lot :smile:
